File: windows/src/Elevate.Core/Providers/AccessPackageProvider.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Elevate.Core.Auth;
using Elevate.Core.Models;
using Elevate.Core.Networking;
using Elevate.Core.Support;

namespace Elevate.Core.Providers;

/// <summary>Self-service entitlement management for the signed-in user. Port of the Swift <c>AccessPackageProvider</c>.</summary>
public interface IAccessPackageProvider
{
    IReadOnlyList<string> Scopes { get; }

    /// <summary>Every package the caller may request, hidden ones included but flagged.</summary>
    Task<IReadOnlyList<AccessPackage>> RequestablePackagesAsync(Identity identity, string tenantId, CancellationToken ct = default);

    Task<IReadOnlyList<AccessPackageRequest>> MyRequestsAsync(Identity identity, string tenantId, CancellationToken ct = default);

    Task<IReadOnlyList<AccessPackageAssignment>> MyAssignmentsAsync(Identity identity, string tenantId, CancellationToken ct = default);

    /// <summary>One entry per policy the caller may request <paramref name="packageId"/> under.</summary>
    Task<IReadOnlyList<PolicyRequirement>> RequirementsAsync(string packageId, Identity identity, string tenantId, CancellationToken ct = default);

    /// <summary>Submits a <c>userAdd</c> request. <paramref name="policyId"/> is required by Graph only when several policies apply.</summary>
    Task<AccessPackageRequest> RequestAsync(string packageId, string? policyId, string justification, Identity identity, string tenantId, CancellationToken ct = default);

    Task CancelAsync(string requestId, Identity identity, string tenantId, CancellationToken ct = default);
}

/// <summary>
/// Self-service entitlement management for the signed-in user: the access packages they may
/// request, their own requests and assignments. Graph v1.0 only.
/// </summary>
public sealed class AccessPackageProvider : IAccessPackageProvider
{
    internal const string Base = "/identityGovernance/entitlementManagement";

    private readonly GraphTransport _transport;

    public AccessPackageProvider(IHttpClient http, ITokenProvider tokens)
        => _transport = new GraphTransport(http, tokens);

    public IReadOnlyList<string> Scopes => Auth.Scopes.EntitlementAll;

    // MARK: Wire shapes

    private sealed record PackageDto(string Id, string? DisplayName, string? Description, bool? IsHidden);

    private sealed record AssignmentRef(string? Id, string? AccessPackageId, string? AssignmentPolicyId);

    private sealed record RequestDto(
        string Id,
        string? RequestType,
        string? State,
        string? Status,
        string? Justification,
        DateTimeOffset? CreatedDateTime,
        DateTimeOffset? CompletedDateTime,
        GraphApprovals.Named? AccessPackage,
        AssignmentRef? Assignment);

    private sealed record AssignmentDto(
        string Id,
        string? State,
        ScheduleInfo? Schedule,
        GraphApprovals.Named? AccessPackage,
        GraphApprovals.Named? AssignmentPolicy);

    /// <summary>Only whether a policy asks questions matters, so the questions themselves are not modelled.</summary>
    private sealed record RequirementDto(
        string? PolicyId,
        string? PolicyDisplayName,
        string? PolicyDescription,
        bool? IsApprovalRequired,
        IReadOnlyList<JsonElement>? Questions);

    private static AccessPackageRequest Request(RequestDto r) => new(
        r.Id,
        r.AccessPackage?.Id ?? r.Assignment?.AccessPackageId ?? string.Empty,
        r.AccessPackage?.DisplayName ?? r.Id,
        r.RequestType ?? "userAdd",
        AccessPackageRequestStates.Parse(r.State),
        r.Status,
        r.Justification,
        r.CreatedDateTime,
        r.CompletedDateTime,
        r.Assignment?.AssignmentPolicyId);

    private static AccessPackageAssignment Assignment(AssignmentDto a) => new(
        a.Id,
        a.AccessPackage?.Id ?? string.Empty,
        a.AccessPackage?.DisplayName ?? a.Id,
        AccessPackageAssignmentStates.Parse(a.State),
        a.AssignmentPolicy?.DisplayName,
        ExpiresAt(a.Schedule));

    /// <summary>An assignment's end, whether Graph sent it as a date or as a duration from the start.</summary>
    private static DateTimeOffset? ExpiresAt(ScheduleInfo? schedule) =>
        schedule?.Expiration is { } expiration && schedule.StartDateTime is { } start
            ? ScheduleRules.End(expiration.EndDateTime, expiration.Duration, start)
            : schedule?.Expiration?.EndDateTime;

    // MARK: Reads

    public async Task<IReadOnlyList<AccessPackage>> RequestablePackagesAsync(Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        var url = _transport.GraphUrl(Base + "/accessPackages/filterByCurrentUser(on='allowedRequestor')");
        var items = await _transport.ListAllAsync<PackageDto>(identity, tenantId, url, Scopes, ct).ConfigureAwait(false);
        return [.. items.Select(p => new AccessPackage(p.Id, p.DisplayName ?? p.Id, p.Description, p.IsHidden ?? false))];
    }

    public async Task<IReadOnlyList<AccessPackageRequest>> MyRequestsAsync(Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        var url = _transport.GraphUrl(Base + "/assignmentRequests/filterByCurrentUser(on='target')?$expand=accessPackage,assignment");
        var items = await _transport.ListAllAsync<RequestDto>(identity, tenantId, url, Scopes, ct).ConfigureAwait(false);
        return [.. items.Select(Request)];
    }

    public async Task<IReadOnlyList<AccessPackageAssignment>> MyAssignmentsAsync(Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        var url = _transport.GraphUrl(Base + "/assignments/filterByCurrentUser(on='target')?$expand=accessPackage,assignmentPolicy");
        var items = await _transport.ListAllAsync<AssignmentDto>(identity, tenantId, url, Scopes, ct).ConfigureAwait(false);
        return [.. items.Select(Assignment)];
    }

    // MARK: Requirements and requests

    public async Task<IReadOnlyList<PolicyRequirement>> RequirementsAsync(string packageId, Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(packageId);
        ArgumentNullException.ThrowIfNull(identity);
        var url = _transport.GraphUrl($"{Base}/accessPackages/{packageId}/getApplicablePolicyRequirements");
        var response = await _transport.PostAsync(identity, tenantId, url, Scopes, [], ct).ConfigureAwait(false);
        var page = JsonSerializer.Deserialize<GraphTransport.Page<RequirementDto>>(response.Body, GraphJson.Options);
        return
        [
            .. (page?.Value ?? [])
                .Where(d => d.PolicyId is not null)
                .Select(d => new PolicyRequirement(
                    d.PolicyId!, d.PolicyDisplayName ?? d.PolicyId!, d.PolicyDescription,
                    d.IsApprovalRequired ?? false, (d.Questions?.Count ?? 0) > 0)),
        ];
    }

    public async Task<AccessPackageRequest> RequestAsync(string packageId, string? policyId, string justification, Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(packageId);
        ArgumentNullException.ThrowIfNull(justification);
        ArgumentNullException.ThrowIfNull(identity);
        var assignment = new JsonObject { ["accessPackageId"] = packageId };
        if (policyId is not null)
        {
            assignment["assignmentPolicyId"] = policyId;
        }

        var body = new JsonObject
        {
            ["requestType"] = "userAdd",
            ["justification"] = justification,
            ["assignment"] = assignment,
        };
        var response = await _transport.PostAsync(
            identity, tenantId, _transport.GraphUrl(Base + "/assignmentRequests"),
            Scopes, Encoding.UTF8.GetBytes(body.ToJsonString()), ct).ConfigureAwait(false);
        var dto = JsonSerializer.Deserialize<RequestDto>(response.Body, GraphJson.Options)
            ?? throw new PimException(PimErrorKind.Unexpected, "Empty response body");
        var created = Request(dto);
        return created.PackageId.Length == 0 ? created with { PackageId = packageId } : created;
    }

    public async Task CancelAsync(string requestId, Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(requestId);
        ArgumentNullException.ThrowIfNull(identity);
        await _transport.PostAsync(
            identity, tenantId, _transport.GraphUrl($"{Base}/assignmentRequests/{requestId}/cancel"), Scopes, [], ct).ConfigureAwait(false);
    }

    /// <summary>The My Access portal page for one package, for policies whose questions Elevate does not collect.</summary>
    public static Uri MyAccessUrl(string tenantId, string packageId) =>
        new($"https://myaccess.microsoft.com/@{tenantId}#/access-packages/{packageId}");
}

File: macos/Sources/ElevateCore/Providers/AccessPackageProvider.swift
import Foundation

/// Self-service entitlement management for the signed-in user: the access packages they may
/// request, their own requests and assignments. Graph v1.0 only.
public struct AccessPackageProvider: Sendable {
    public let scopes = EntitlementScopes.all
    let transport: GraphTransport

    public init(http: any HTTPClient, tokens: any TokenProviding) {
        transport = GraphTransport(http: http, tokens: tokens)
    }

    static let base = "/identityGovernance/entitlementManagement"

    // MARK: Wire shapes

    struct Named: Decodable { let id: String?; let displayName: String? }
    struct PackageDTO: Decodable { let id: String; let displayName: String?; let description: String?; let isHidden: Bool? }
    struct AssignmentRef: Decodable { let id: String?; let accessPackageId: String?; let assignmentPolicyId: String? }
    struct RequestDTO: Decodable {
        let id: String
        let requestType: String?
        let state: String?
        let status: String?
        let justification: String?
        let createdDateTime: Date?
        let completedDateTime: Date?
        let accessPackage: Named?
        let assignment: AssignmentRef?
    }
    struct Expiration: Decodable { let type: String?; let endDateTime: Date? }
    struct Schedule: Decodable { let startDateTime: Date?; let expiration: Expiration? }
    struct AssignmentDTO: Decodable {
        let id: String
        let state: String?
        let status: String?
        let schedule: Schedule?
        let accessPackage: Named?
        let assignmentPolicy: Named?
    }

    static func request(from r: RequestDTO) -> AccessPackageRequest {
        AccessPackageRequest(id: r.id,
                             packageId: r.accessPackage?.id ?? r.assignment?.accessPackageId ?? "",
                             packageName: r.accessPackage?.displayName ?? r.id,
                             requestType: r.requestType ?? "userAdd",
                             state: AccessPackageRequestState.parse(r.state),
                             status: r.status, justification: r.justification,
                             createdAt: r.createdDateTime, completedAt: r.completedDateTime,
                             policyId: r.assignment?.assignmentPolicyId)
    }

    static func assignment(from a: AssignmentDTO) -> AccessPackageAssignment {
        AccessPackageAssignment(id: a.id,
                                packageId: a.accessPackage?.id ?? "",
                                packageName: a.accessPackage?.displayName ?? a.id,
                                state: AccessPackageAssignmentState.parse(a.state),
                                policyName: a.assignmentPolicy?.displayName,
                                expiresAt: a.schedule?.expiration?.endDateTime)
    }

    // MARK: Reads

    public func requestablePackages(identity: Identity, tenantId: String) async throws -> [AccessPackage] {
        let url = try transport.graphURL("\(Self.base)/accessPackages/filterByCurrentUser(on='allowedRequestor')")
        let items = try await transport.listAll(PackageDTO.self, identity: identity, tenantId: tenantId, url: url, scopes: scopes)
        return items.map { AccessPackage(id: $0.id, displayName: $0.displayName ?? $0.id, description: $0.description, isHidden: $0.isHidden ?? false) }
    }

    public func myRequests(identity: Identity, tenantId: String) async throws -> [AccessPackageRequest] {
        let url = try transport.graphURL("\(Self.base)/assignmentRequests/filterByCurrentUser(on='target')?$expand=accessPackage,assignment")
        let items = try await transport.listAll(RequestDTO.self, identity: identity, tenantId: tenantId, url: url, scopes: scopes)
        return items.map(Self.request(from:))
    }

    public func myAssignments(identity: Identity, tenantId: String) async throws -> [AccessPackageAssignment] {
        let url = try transport.graphURL("\(Self.base)/assignments/filterByCurrentUser(on='target')?$expand=accessPackage,assignmentPolicy")
        let items = try await transport.listAll(AssignmentDTO.self, identity: identity, tenantId: tenantId, url: url, scopes: scopes)
        return items.map(Self.assignment(from:))
    }
}
